src/components/TrapFocus/TrapFocus.tsx

```diff
@@ -13,36 +13,27 @@ import {
 } from '../../utilities/focus';
 import {useFocusManager} from '../../utilities/focus-manager';
 import {portal} from '../shared';
+import {useIsAfterInitialMount} from '../../utilities/use-is-after-initial-mount';
 
 export interface TrapFocusProps {
   trapping?: boolean;
   children?: React.ReactNode;
 }
 
 export function TrapFocus({trapping = true, children}: TrapFocusProps) {
-  const [shouldFocusSelf, setFocusSelf] = useState<boolean | undefined>(
-    undefined,
-  );
   const {canSafelyFocus} = useFocusManager({trapping});
   const focusTrapWrapper = useRef<HTMLDivElement>(null);
+  const isMounted = useIsAfterInitialMount();
 
-  useEffect(() => {
-    setFocusSelf(
-      !(
-        canSafelyFocus &&
-        focusTrapWrapper.current &&
-        focusTrapWrapper.current.contains(document.activeElement)
-      ),
-    );
-  }, [canSafelyFocus]);
-
-  const shouldDisableFirstElementFocus = () => {
-    if (shouldFocusSelf === undefined || !canSafelyFocus) {
-      return true;
-    }
-
-    return shouldFocusSelf ? !trapping : !shouldFocusSelf;
-  };
+  const disableFocus =
+    isMounted &&
+    canSafelyFocus &&
+    !(
+      focusTrapWrapper.current &&
+      focusTrapWrapper.current.contains(document.activeElement)
+    )
+      ? !trapping
+      : true;
```

Comment:
Simplifying the logic ironed out the kinks. We're saying if the component is not mounted, not the main trap focus target (e.g. multiple trap focuses active), or focus is within the component disable `Focus` other wise following the `trapping` prop.